src/term/transform/scott_generation.rs

```diff
@@ -0,0 +1,84 @@
+use crate::term::{Book, Definition, Name, Rule, Term};
+
+impl Book {
+  pub fn generate_scott_adts(&mut self) {
+    for adt in self.adts.values() {
+      let curr_ctrs = adt.ctrs.clone();
+
+      let mut ctr_args = Vec::new();
+      let mut ctrs = Vec::new();
+
+      for (ctr_name, arity) in &curr_ctrs {
+        for ctr in curr_ctrs.keys() {
+          ctrs.push(ctr.clone());
+          if ctr == ctr_name {
+            make_args(arity, &mut ctr_args);
+          }
+        }
+
+        let lam = make_lam(&ctr_args, &mut ctrs, ctr_name);
+
+        ctr_args.clear();
+        ctrs.clear();
+
+        let def_id = self.def_names.insert(ctr_name.clone());
+
+        let rule = Rule { def_id, pats: vec![], body: lam };
```

Comment:
Consider inlining these two variables `rule` and `rules`. They are pretty straight forward and I don't think we gain by separating these lines.

Reply:
Maybe just one of them:
`let rules = vec![Rule{...}];`
or
```
let rule = Rule {...};
let def = Definition {def_id, rules: vec![rule]};
```